### PR TITLE
add NewReplyFromDHCPv6Message

### PR DESCRIPTION
**Pull request by pmazzini**
Add a more generic NewReplyFromDHCPv6Message to avoid code duplication.
This is one step closer towards fixing issue #73.

### CODE REVIEW

File: dhcpv6/dhcpv6message.go [hunk 1]
```diff
@@ -198,26 +198,32 @@ func NewRequestFromAdvertise(advertise DHCPv6, modifiers ...Modifier) (DHCPv6, e
...
-	if rebind.Type() != REBIND {
-		return nil, errors.New("The passed REBIND must have REBIND type set")
-	}
-	reb, ok := rebind.(*DHCPv6Message)
+// NewReplyFromMessage creates a new REPLY packet based on a MESSAGE packet.
+func NewReplyFromMessage(message DHCPv6, modifiers ...Modifier) (DHCPv6, error) {
+	if message == nil {
+		return nil, errors.New("MESSAGE cannot be nil")
+	}
+	switch message.Type() {
```
Comment (insomniacslk) [inline dhcpv6/dhcpv6message.go]: you can put all on a single line:
```
switch message.Type() {
    case REQUEST, CONFIRM, RENEW, REBIND, RELEASE:
        ....
}
```

File: dhcpv6/dhcpv6message.go [hunk 2]
```diff
@@ -198,26 +198,32 @@ func NewRequestFromAdvertise(advertise DHCPv6, modifiers ...Modifier) (DHCPv6, e
...
-// NewReplyFromRebind creates a new REPLY packet based on a REBIND packet.
-func NewReplyFromRebind(rebind DHCPv6, modifiers ...Modifier) (DHCPv6, error) {
-	if rebind == nil {
-		return nil, errors.New("REBIND cannot be nil")
-	}
-	if rebind.Type() != REBIND {
-		return nil, errors.New("The passed REBIND must have REBIND type set")
-	}
-	reb, ok := rebind.(*DHCPv6Message)
+// NewReplyFromMessage creates a new REPLY packet based on a MESSAGE packet.
```
Comment (insomniacslk) [inline dhcpv6/dhcpv6message.go]: s/Message/DHCPv6Message/ is that what you mean? There is no MESSAGE packet type, so that would look a bit misleading. Alternatively you can mention explicitly all the message types that are considered for creating the REPLY message (i.e. solicit, renew, etc)

File: dhcpv6/dhcpv6message.go [hunk 3]
```diff
@@ -198,26 +198,32 @@ func NewRequestFromAdvertise(advertise DHCPv6, modifiers ...Modifier) (DHCPv6, e
...
+func NewReplyFromMessage(message DHCPv6, modifiers ...Modifier) (DHCPv6, error) {
+	if message == nil {
+		return nil, errors.New("MESSAGE cannot be nil")
```
Comment (insomniacslk) [inline dhcpv6/dhcpv6message.go]: same as above

File: dhcpv6/dhcpv6message.go [hunk 4]
```diff
@@ -229,78 +235,33 @@ func NewReplyFromRebind(rebind DHCPv6, modifiers ...Modifier) (DHCPv6, error) {
...
 
-// NewReplyFromRenew creates a new REPLY packet based on a RENEW packet.
-func NewReplyFromRenew(renew DHCPv6, modifiers ...Modifier) (DHCPv6, error) {
-	if renew == nil {
-		return nil, errors.New("RENEW cannot be nil")
-	}
-	if renew.Type() != RENEW {
-		return nil, errors.New("The passed RENEW must have RENEW type set")
+// NewReplyFromMessageWithServerID creates a new REPLY packet based on a MESSAGE packet.
+func NewReplyFromMessageWithServerID(message DHCPv6, modifiers ...Modifier) (DHCPv6, error) {
```
Comment (insomniacslk) [inline dhcpv6/dhcpv6message.go]: I think this function is not necessary and should be rather done by calling `NewReplyFromMessage` with a modifier, i.e.
```
reply := NewReplyFromMessage(msg, WithServerID(msg.GetOneOption(OPTION_SERVER_ID)))
```

I would leave to the user to check that the ServerID contained in msg is not `nil`

**Comment by insomniacslk** [comment]
did you remove NewReplyFrom{Request,Rebind,Renew}? I assume this is by mistake?

**Comment by pmazzini** [comment]
> did you remove NewReplyFrom{Request,Rebind,Renew}? I assume this is by mistake?
No, I think they are useless after adding NewReplyFromDHCPv6Message.
Do we want to maintain backwards compatibility?

**Comment by pmazzini** [comment]
I am not sure why the build job for go 1.10 failed. Can you trigger a new one to see if it is something transient?

**Comment by insomniacslk** [comment]
> No, I think they are useless after adding NewReplyFromDHCPv6Message.
> Do we want to maintain backwards compatibility?

No, I think it's good to remove them, we haven't stabilized the API yet

File: dhcpv6/dhcpv6message.go [hunk 5]
```diff
@@ -198,26 +198,28 @@ func NewRequestFromAdvertise(advertise DHCPv6, modifiers ...Modifier) (DHCPv6, e
 	return d, nil
 }
 
-// NewReplyFromRebind creates a new REPLY packet based on a REBIND packet.
-func NewReplyFromRebind(rebind DHCPv6, modifiers ...Modifier) (DHCPv6, error) {
-	if rebind == nil {
-		return nil, errors.New("REBIND cannot be nil")
+// NewReplyFromDHCPv6Message creates a new REPLY packet based on a DHCPv6Message.
```
Comment (insomniacslk) [inline dhcpv6/dhcpv6message.go]: Please add some more information in the comment, in particular that this function is to be used when generating a reply to REQUEST, CONFIRM, RENEW, REBIND and RELEASE

File: dhcpv6/dhcpv6_test.go [hunk 6]
```diff
@@ -82,72 +82,39 @@ func TestFromAndToBytes(t *testing.T) {
...
+	a, err := NewAdvertiseFromSolicit(&s, WithServerID(duid))
 	require.NoError(t, err)
 	require.Equal(t, a.(*DHCPv6Message).TransactionID(), s.TransactionID())
 	require.Equal(t, a.Type(), ADVERTISE)
 }
 
-func TestNewReplyFromRequest(t *testing.T) {
+func TestNewReplyFromDHCPv6Message(t *testing.T) {
 	req := DHCPv6Message{}
 	req.SetMessage(REQUEST)
```
Comment (insomniacslk) [inline dhcpv6/dhcpv6_test.go]: please also test the other cases (CONFIRM, RENEW, REBIND and RELEASE), and the negative test cases (e.g. SOLICIT and others)

**Comment by insomniacslk** [comment]
Thanks a lot for this PR and for your patience in making the requested changes!
